### ISSUE TITLE
Support for generative models T5, BERT generation, BART

### ISSUE THREAD

**Issue by AlexMikhalev** [RedisAI/RedisAI#672]
There is a number of models which in addition to forward() method have a generate() method, simplest example using T5 for text summarisation:
```
text = """At a very high level, one of the most critical steps in any ML pipeline is called AI serving, a task usually performed by an AI inference engine. The AI inference engine is responsible for the model deployment and performance monitoring steps in the figure above, and represents a whole new world that will eventually determine whether applications can use AI technologies to improve operational efficiencies and solve real business problems. """
text
from transformers import T5ForConditionalGeneration, T5Tokenizer
model = T5ForConditionalGeneration.from_pretrained("t5-base")
tokenizer = T5Tokenizer.from_pretrained("t5-base")
tokenizer
inputs = tokenizer.encode("summarize: " + text, return_tensors="pt", max_length=512, truncation=True)
inputs
outputs = model.generate(
    inputs, 
    max_length=150, 
    min_length=40, 
    length_penalty=2.0, 
    num_beams=4, 
    early_stopping=True)
# just for debugging
print(outputs)
print(tokenizer.decode(outputs[0]))
```
We had a discussion with @DvirDukhan in chat and tried to redefine generate via forward class:

```
In [26]: class MyModel(T5ForConditionalGeneration):
    ...:   def __init__(self,kwargs):
    ...:     super(MyModel, self).__init__(kwargs) # could be other init params
    ...: 
    ...:   def forward(self, params):
    ...:     self.generate(params)
    ...: 

In [27]: model1=MyModel.from_pretrained("t5-base", torchscript=True)

In [28]: outputs1 = model1(inputs)
```
but generate function calls forward internally. So far I could not find a way to trace Generative model and load it into Redis AI.
My understanding Transformers implementation inherits generate from [this class](https://github.com/huggingface/transformers/blob/master/src/transformers/generation_utils.py#L665), with fairly large list of models using generate. 

Ideas or suggestions?

**Comment by DvirDukhan** [comment]
hi @AlexMikhalev 
I had a discussion with @lantiga 
the solution might be here
https://huggingface.co/transformers/serialization.html
Where you can export to ONNX or PyTorch models.
note:
If the model is too big try to quantize the model, as protobuf limitation is 2 GB.

**Comment by AlexMikhalev** [comment]
Hello @DvirDukhan and @lantiga, 
I am not sure I understand the directions: 
With your help, I can save and load standard BERT models (and yes, my protobuf 2GB): [export and load](https://github.com/applied-knowledge-systems/the-pattern-api/blob/12c2a1e79b93a48366cb02dfa4e4bfd5b08a7cf5/qasearch/export_load_bert.py), it's the fact that generative models have forward and backwards pass inside the separate model.generate() call which is puzzling. 
Calling BERT QA is [fine](https://github.com/applied-knowledge-systems/the-pattern-api/blob/12c2a1e79b93a48366cb02dfa4e4bfd5b08a7cf5/qasearch/qa_bert.py#L46). 
Does ONNX RedisAI API have a different mechanism to allow multiple methods from model, not only forward()?

**Comment by DvirDukhan** [comment]
@AlexMikhalev 
As far as I recall @lantiga (and feel free to correct me if I'm wrong) explenation, the `generate()` function calls many time (in your case: between 40 to 150 times) to `forward` in order to generate an output (or until stop). The `backwards` is far is I recall is for computing gradients in the learning phase.
With regards to your original question, can you please check this issue? https://github.com/huggingface/transformers/issues/8923

**Comment by AlexMikhalev** [comment]
Hello @DvirDukhan, yes I have seen this bug. I am not sure what would be the best way forward - most examples with generative text models lead to transformers (although I managed to find a few with PyTorch coded from scratch). 
I am currently exploring how https://github.com/LaurentMazare/tch-rs binds to PyTorch so they can expose calls of pre-trained models into Rust (rust-bert).

**Comment by AlexMikhalev** [comment]
I found another implementation of BART summarisation: 
https://github.com/pytorch/fairseq/tree/master/examples/bart which have a simple enough generate() function in [summarize.py](https://github.com/pytorch/fairseq/blob/master/examples/bart/summarize.py) I may try to turn it to TorchScript to run inside RedisAI, but it will open question to pass hashtag() to touch script. 
Thoughts?

**Comment by lantiga** [comment]
Hello @AlexMikhalev, thanks for raising the issue.
At the present stage, beam search or greedy approaches for transformers are beyond the reach of RedisAI. The reason is that currently you cannot call MODELRUN from a SCRIPT, so you would need to run beam search or greedy strategies from the client, calling into RedisAI to run individual encoder/decoder calls.

However, we are looking into this very closely. In particular, we are planning to enable calling MODELRUN from SCRIPT, which would open up the possibility to provide a series of standard TorchScript scripts to make it possible to do what you are looking for. We'll keep you posted!

**Comment by OwlWisdom** [comment]
I'm struggling with the same problem and I'm trying to solve it but I don't feel capable of doing so.
